File: algorithm/js/list_scheduling.js
function findMinimunIndex(array) {
  let min = array[0];
  let minIndex = 0;
  for (let i = 0; i < array.length; i++) {
    if (array[i] < array[minIndex]) {
      min = array[i];
      minIndex = i;
    }
  }

  return minIndex;
}

function listScheduling(tasks, n_workers, n_tasks) {
  var workers = Array(n_workers).fill([]);
  var current_time_workers = Array(n_workers).fill(0);

  for (let i = 0; i < n_workers; i++) {
    workers[i] = Array(n_tasks).fill(0);
  }

  for (var i = n_tasks - 1; i >= 0; i--) {
    var minimunIndex = findMinimunIndex(current_time_workers);
    current_time_workers[minimunIndex] += tasks[i];
    const tasksIndex = findMinimunIndex(workers[minimunIndex]);
    workers[minimunIndex][tasksIndex] = tasks[i];
  }

  return workers;
}

module.exports = listScheduling;

File: algorithm/c/list_scheduling.c
#include <stdlib.h>
#include "mergesort.h"

int findMinimunIndex(int* array, int lenArray)
{
  int min = array[0];
  int index = 0;
  for (int i = 1; i < lenArray; i++)
  {
    if (array[i] < min)
    {
      min = array[i];
      index = i;
    }
  }

  return index;
}

int** listScheduling(int* tasks, int n_workers, int n_tasks)
{
  int* current_time_workers = calloc(n_workers, sizeof(int));
  int** workers = calloc(n_workers, sizeof(int*));
  mergeSort(tasks, 0, n_tasks - 1);
  for (int i = 0; i < n_workers; i++)
  {
    workers[i] = calloc(n_tasks, sizeof(int));
  }
  
  for (int i = n_tasks - 1; i >= 0; i--)
  {
    int minimunIndex = findMinimunIndex(current_time_workers, n_workers);
    current_time_workers[minimunIndex] += tasks[i];
    int taskIndex = findMinimunIndex(workers[minimunIndex], n_tasks);
    workers[minimunIndex][taskIndex] = tasks[i];
  }
  free(current_time_workers);

  return workers;
}
